File: day12/day12.go
package main

import (
	"fmt"
	"strconv"

	aocutils "../common"
)

func main() {
	input, _ := aocutils.GetInputStrings(aocutils.FindInputFile(12))

	fmt.Println("Part 1: ", part1(input))
	fmt.Println("Part 2: ", part2(input))
}

func part1(input []string) int {
	x, y, face := 0, 0, 90
	for _, direction := range input {
		d := direction[0]
		count, _ := strconv.Atoi(direction[1:])

		switch d {
		case 'N':
			y += count
		case 'S':
			y -= count
		case 'W':
			x -= count
		case 'E':
			x += count
		case 'L':
			face = (360 + face - count) % 360
		case 'R':
			face = (face + count) % 360
		case 'F':
			if face == 0 {
				y += count
			} else if face == 90 {
				x += count
			} else if face == 180 {
				y -= count
			} else if face == 270 {
				x -= count
			}
		}
	}
	return aocutils.AbsInt(x) + aocutils.AbsInt(y)
}

func part2(input []string) int {
	wX, wY, sX, sY := 10, 1, 0, 0
	for _, direction := range input {
		d := direction[0]
		count, _ := strconv.Atoi(direction[1:])

		switch d {
		case 'N':
			wY += count
		case 'S':
			wY -= count
		case 'W':
			wX -= count
		case 'E':
			wX += count
		case 'L':
			wX, wY = rotate(wX, wY, sX, sY, count)
		case 'R':
			wX, wY = rotate(wX, wY, sX, sY, 360-count)
		case 'F':
			// Get waypoint distance from ship
			dx := wX - sX
			dy := wY - sY
			// Move ship 'count' times in direction of waypoint
			sX += (dx * count)
			sY += (dy * count)
			// Maintain waypoint distance from ship
			wX = sX + dx
			wY = sY + dy
		}
	}
	return aocutils.AbsInt(sX) + aocutils.AbsInt(sY)
}

// rotate clockwise through 0, 90, 180 or 270 around a non-zero origin
func rotate(pX, pY, originX, originY, angle int) (int, int) {
	offX, offY := pX-originX, pY-originY
	switch angle {
	case 0:
		return pX, pY
	case 90:
		return -offY + originX, offX + originY
	case 180:
		return -offX + originX, -offY + originY
	case 270:
		return offY + originX, -offX + originY
	default:
		panic("Unexpected angle to rotate")
	}
}

File: day12/day12.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.Extensions.Configuration;
using Serilog;

namespace Shunty.AdventOfCode2020
{
    public class Day12 : AoCRunnerBase
    {
        public override async Task Execute(IConfiguration config, ILogger logger, bool useTestData)
        {
            var input = await GetInputLines(useTestData);

            ShowResult(1, Part1(input));
            ShowResult(2, Part2(input));
        }

        private int Part1(IList<string> input)
        {
            int x = 0, y = 0, facing = 90;
            foreach (var dir in input)
            {
                if (string.IsNullOrWhiteSpace(dir))
                    continue;

                var d = dir[0];
                var i = int.Parse(dir.Substring(1));
                switch (d)
                {
                    case 'N':
                        y += i;
                        break;
                    case 'S':
                        y -= i;
                        break;
                    case 'W':
                        x -= i;
                        break;
                    case 'E':
                        x += i;
                        break;
                    case 'L':
                        facing = (facing + 360 - i) % 360;
                        break;
                    case 'R':
                        facing = (facing + i) % 360;
                        break;
                    case 'F':
                        switch (facing)
                        {
                            case 0:
                                y += i;
                                break;
                            case 90:
                                x += i;
                                break;
                            case 180:
                                y -= i;
                                break;
                            case 270:
                                x -= i;
                                break;
                            default:
                                throw new Exception($"Don't know how to handle angle of {facing} degrees");
                        }
                        break;
                    default:
                        throw new Exception($"Unknown direction command {d}");
                }
            }
            return Math.Abs(x) + Math.Abs(y);
        }

        private Int64 Part2(IList<string> input)
        {
            Int64 x = 0, y = 0, wayX = 10, wayY = 1;
            foreach (var dir in input)
            {
                if (string.IsNullOrWhiteSpace(dir))
                    continue;

                var d = dir[0];
                var i = int.Parse(dir.Substring(1));
                switch (d)
                {
                    case 'N':
                        wayY += i;
                        break;
                    case 'S':
                        wayY -= i;
                        break;
                    case 'W':
                        wayX -= i;
                        break;
                    case 'E':
                        wayX += i;
                        break;
                    case 'L':
                        (wayX, wayY) = Rotate(wayX, wayY, x, y, i);
                        break;
                    case 'R':
                        (wayX, wayY) = Rotate(wayX, wayY, x, y, 360 - i);
                        break;
                    case 'F':
                        // Move 'i' times in the direction of the waypoint
                        Int64 dx = wayX - x, dy = wayY - y;
                        x += (dx * i);
                        y += (dy * i);
                        // Move the waypoint to maintain its distance from the ship
                        wayX = x + dx;
                        wayY = y + dy;
                        break;
                    default:
                        throw new Exception($"Unknown direction command {d}");
                }
            }
            return Math.Abs(x) + Math.Abs(y);
        }

        // Rotate the waypoint about the ship
        private (Int64,Int64) Rotate(Int64 wX, Int64 wY, Int64 sX, Int64 sY, int angle)
        {
            // Get waypoint co-ords relative to the ship
            Int64 x = wX - sX, y = wY - sY;
            // Only deal with 0, 90, 180,270
            (Int64 X, Int64 Y) result;
            switch (angle)
            {
                case 0:
                case 360:
                    result = (x,y);
                    break;
                case 90:
                    result = (-y,x);
                    break;
                case 180:
                    result = (-x,-y);
                    break;
                case 270:
                    result = (y,-x);
                    break;
                default:
                    throw new Exception($"Unhandled rotation for angle {angle}");
            }
            // Add the ship co-ords back in
            return (result.X + sX, result.Y + sY);
        }
    }
}
